### PR TITLE
 Digital camera VideoCapture - connection to gPhoto2 library

### PR DESCRIPTION
**Pull request by dobrypd**
I've created connection to [gPhoto2](http://www.gphoto.org/) library.
It supports plenty of digital cameras ([list](http://www.gphoto.org/proj/libgphoto2/support.php)) for direct capture, liveview and configuration.
Example of usage - implementation of [autofocus](https://github.com/dobrypd/autofocus).

### CODE REVIEW

**Comment by StevenPuttemans** [comment]
Hi there! It would be cool if you could add the autofocus example that you are linking as a sample to the repository also, to have a sample of how to use the interface. As to the reason why there are still build errors, `modules/videoio/src/cap_gphoto2.cpp:1219: new blank line at EOF.` so simply remove the last blank line!

**Comment by dobrypd** [comment]
Hi Steven! Thanks for the response.
I've added the example and removed last line.
I'm willing to hear for any other idea for change or improvement.

File: samples/cpp/autofocus.cpp [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+    // Update state.
+    state.rate = rate;
+    state.stepToLastMax -= state.direction * state.step;
+    return state.step;
+}
+
+static void showHelp(const char * pName, bool welcomeMsg)
+{
+    cout
+            << "This program demonstrates usage of gPhoto2 VideoCapture.\n\n"
```
Comment (StevenPuttemans) [inline samples/cpp/autofocus.cpp]: No clue why but identation seems a bit lost here, could you fix?

File: samples/cpp/autofocus.cpp [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+                    "With OpenCV build without gPhoto2 library support it will "
+                    "do nothing special, just capture.\n\n"
+                    "Simple implementation of autofocus is based on edges detection.\n"
+                    "It was tested (this example) only with Nikon DSLR (Nikon D90).\n"
+                    "But shall work on all Nikon DSLRs, and with little effort with other devices.\n"
+                    "Visit http://www.gphoto.org/proj/libgphoto2/support.php\n"
+                    "to find supported devices (need Image Capture at least).\n"
+                    "Before run, set your camera autofocus ON.\n\n";
+    if (!welcomeMsg)
+        cout << "usage " << pName << ": [OPTIONS] DEVICE_NAME\n\n"
```
Comment (StevenPuttemans) [inline samples/cpp/autofocus.cpp]: There should be brackets around the complete if statement

File: samples/cpp/autofocus.cpp [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+                "\t-f FPS\t\tframes per second in output video,\n"
+                "\t-m\t\tmeasure exposition\n"
+                "\t\t\t(returns rates from closest focus to INTY\n"
+                "\t\t\tfor every minimum step),\n"
+                "\t-d INT\t\tset minimum focus step,\n"
+                "\t-v\t\tverbose mode.\n\n\n"
+                "DEVICE_NAME\t\tis your digital camera model substring.\n\n\n"
+                "On runtime you can use keys to control:\n";
+    if (welcomeMsg)
+        cout << "Actions:\n";
```
Comment (StevenPuttemans) [inline samples/cpp/autofocus.cpp]: Here also + identation fix

File: samples/cpp/autofocus.cpp [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+    FocusState state = createInitialState();
+    bool focus = true;
+    bool lastSucceeded = true;
+    namedWindow(windowOriginal, 1);
+
+    // Get settings:
+    if (GlobalArgs.verbose)
+    {
+        cout << "List of camera settings: " << endl
+                << (const char *) (intptr_t) cap.get(
```
Comment (StevenPuttemans) [inline samples/cpp/autofocus.cpp]: Can you make identation equal here?

File: samples/cpp/autofocus.cpp [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+        cap.set(CAP_PROP_GPHOTO2_COLLECT_MSGS, true);
+    }
+
+    cap.set(CAP_PROP_GPHOTO2_PREVIEW, true);
+    cap.set(CAP_PROP_VIEWFINDER, true);
+    cap >> frame; // To check PREVIEW output Size.
+    if (GlobalArgs.output != NULL)
+    {
+        Size S = Size((int) cap.get(CAP_PROP_FRAME_WIDTH),
+                (int) cap.get(CAP_PROP_FRAME_HEIGHT));
```
Comment (StevenPuttemans) [inline samples/cpp/autofocus.cpp]: back to precious line?

File: samples/cpp/autofocus.cpp [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+        videoWriter.open(GlobalArgs.output, CV_FOURCC('M', 'J', 'P', 'G'),
+                GlobalArgs.fps, S, true);
```
Comment (StevenPuttemans) [inline samples/cpp/autofocus.cpp]: back to previous line?

File: samples/cpp/autofocus.cpp [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+        }
+    }
+    showHelp(argv[0], true); // welcome msg
+
+    if (GlobalArgs.minimumFocusStep == 0)
+        state.minFocusStep = findMinFocusStep(cap, FOCUS_STEP / 16,
+                -FOCUS_DIRECTION_INFTY);
+    else
+        state.minFocusStep = GlobalArgs.minimumFocusStep;
+    focusDriveEnd(cap, -FOCUS_DIRECTION_INFTY); // Start with closest
```
Comment (StevenPuttemans) [inline samples/cpp/autofocus.cpp]: not clear if this has to be in the else statement or not, else use brackets!

**Comment by StevenPuttemans** [comment]
Okay most comments I made are pure identation fixes or adding brackets for clarity! I do know that it is not always neccesary, BUT it can help the readability of your code. 

For the rest, seems good now. Last thing you should do is squash commits and make a universal commit message. This can be done by:

   1.  Go to your console and browse to your repo
   2. Type git rebase -i HEAD~4 (or accordingly to more fixes)
   3. Rename all picks except for the first one to squash, do not temper with anything else, save
   4. Add a single message for the changes, your first will probably be best
   5. Save and commit
   6. Now do a git push -f --all origin and your change will be made

**Comment by dobrypd** [comment]
Thanks for the review! Changes are applied.

File: CMakeLists.txt [hunk 8]
```diff
@@ -1014,6 +1015,10 @@ if(DEFINED WITH_INTELPERC)
   status("    Intel PerC:"     HAVE_INTELPERC      THEN "YES"                                 ELSE NO)
 endif(DEFINED WITH_INTELPERC)
 
+if(DEFINED WITH_GPHOTO2)
+  status("    gPhoto2:"          HAVE_GPHOTO2          THEN "YES"                                        ELSE NO)
```
Comment (StevenPuttemans) [inline CMakeLists.txt]: Could you try aligning the `HAVE_GPHOTO2` with the above `HAVE_INTELPERC`, this is just a beauty fix but one many programmers like. Same for the other parts of that line.

File: modules/videoio/include/opencv2/videoio.hpp [hunk 9]
```diff
@@ -380,6 +381,24 @@ enum { VIDEOWRITER_PROP_QUALITY = 1,    // Quality (0..100%) of the videostream
        VIDEOWRITER_PROP_FRAMEBYTES = 2, // (Read-only): Size of just encoded video frame
      };
 
+// gPhoto2 properties, if propertyId is less than 0 then work on widget with that __additive inversed__ camera setting ID
+// Get IDs by using CAP_PROP_GPHOTO2_WIDGET_ENUMERATE.
+// @see CvCaptureCAM_GPHOTO2 for more info
+enum {
```
Comment (StevenPuttemans) [inline modules/videoio/include/opencv2/videoio.hpp]: A good practice is to align all the comments for each parameter below at the same identation. Could you apply this? Again beauty fix for clarification.

File: modules/videoio/include/opencv2/videoio/videoio_c.h [hunk 10]
```diff
@@ -391,6 +393,22 @@ enum
     CV_CAP_INTELPERC_IMAGE                  = 3
 };
 
+// gPhoto2 properties, if propertyId is less than 0 then work on widget with that __additive inversed__ camera setting ID
+// Get IDs by using CAP_PROP_GPHOTO2_WIDGET_ENUMERATE.
+// @see CvCaptureCAM_GPHOTO2 for more info
+enum {
```
Comment (StevenPuttemans) [inline modules/videoio/include/opencv2/videoio/videoio_c.h]: Same here. However I am not sure that you should make the fixes to this _c header. It is mainly there for backwards compatibility but as far as I know we do not add new functionality in the C interface. @vpisarev could you clarify this?

File: modules/videoio/src/cap_gphoto2.cpp [hunk 11]
```diff
@@ -0,0 +1,1218 @@
...
+/**
+ * @param output will be changed if possible, return 0 if changed,
+ * @return widget, or NULL if output value was found (saved in argument),
+ */
+CameraWidget * DigitalCameraCapture::getGenericProperty(int propertyId,
+        double & output) const
+{
+    switch (propertyId)
+    {
+    case CV_CAP_PROP_POS_MSEC:
```
Comment (StevenPuttemans) [inline modules/videoio/src/cap_gphoto2.cpp]: The cases below do not have a correct identation. They should be 1 further than the switch statement. Could you fix?

**Comment by vpisarev** [comment]
@dobrypd, looks like there are merge conflicts because of recent patches that also add new functionality to videoio headers. Could you rebase your patch and I will merge it in

**Comment by dobrypd** [comment]
Thanks a lot for the last review.
I'm finally back. Today I've applied fixes and merge. Please do not merge, tomorrow I'll review today's changes and do rebase (marge commits).

**Comment by dobrypd** [comment]
Well, that list of commits after rebased merge looks disgusting.
If it's too much remove PR and I'll create a new one.

**Comment by bhack** [comment]
Rebasing generally it is enough. But now this it is closed and the new PR is at https://github.com/Itseez/opencv/pull/4040

**Pull request closed without merging**